XenonCode.hpp

```diff
@@ -2742,41 +2763,48 @@ const int VERSION_PATCH = 0;
 			auto openFunction = [&](const std::string& name){
 				assert(currentScope == 0);
 				assert(currentFunctionName == "");
+				ram_numericVariables_offset = ram_numericVariables;
+				ram_textVariables_offset = ram_textVariables;
+				ram_objectReferences_offset = ram_objectReferences;
+				ram_numericArrays_offset = ram_numericArrays;
+				ram_textArrays_offset = ram_textArrays;
+		
 				if (name != "system.timer" && name != "system.input" && !name.starts_with("entrypoint.") && functionRefs.contains(name)) {
 					throw CompileError("Function " + name + " is already defined");
 				}
 				currentFunctionName = name;
 				currentFunctionAddr = addr();
+				if (currentFunctionName == "system.timer") {
+					assert(currentTimerInterval);
+					timers.emplace_back(currentTimerInterval, currentFunctionAddr);
+					currentTimerInterval = 0;
+					userVars["system.timer"].clear();
+				} else if (currentFunctionName == "system.input") {
+					inputs[currentInputPort].addr = currentFunctionAddr;
+					currentInputPort = 0;
+					userVars["system.input"].clear();
+				} else if (currentFunctionName.starts_with("entrypoint.")) {
+					assert(entryPoints.size() > 0);
+					entryPoints.back().addr = currentFunctionAddr;
+					userVars[currentFunctionName].clear();
+				} else if (currentFunctionName != "") {
+					functionRefs.emplace(currentFunctionName, currentFunctionAddr);
```

Comment:
This is bad, it removes the anti-recursion protection for all functions.. It must stay in the `closeCurrentFunction` scope. 
Also because of the `self` keyword, you still don't need the recursive functions to be defined before they're out of scope. 
I don't think this block should have been moved at all.